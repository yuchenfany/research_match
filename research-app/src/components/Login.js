/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import '../assets/index.css';

function Login({ user, setUser }) {
  const [error, setError] = useState({ message: '' });
  const navigate = useNavigate();
  // const [samePassword, setSamePassword] = useState(0);

  async function handleSubmit(event) {
    // console.log(user.username);
    // console.log(user.password);

    if (user.username.length === 0 && user.password.length === 0) {
      setError({ message: 'Please enter your login credentials' });
      event.preventDefault();
      return;
    }
    if (user.username.length === 0) {
      setError({ message: 'Please enter your username' });
      event.preventDefault();
      return;
    }
    if (user.password.length === 0) {
      setError({ message: 'Please enter your password' });
      event.preventDefault();
      return;
    }

    const data = await fetch(`http://localhost:5000/record/${user.username}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    const json = await data.json();

    // verification checks of username & password
    if (json === null) {
      setError({ message: 'User does not exist' });
      event.preventDefault();
    } else if (json.password === user.password) {
      navigate('/participant-home');
    } else {
      console.log('INCORRECT PASSWORD');
      setError({ message: 'Incorrect password' });
      event.preventDefault();
    }
  }

  // update username as it's being entered
  const handleNameChange = async (event) => {
    setUser(
      {
        username: event.target.value,
        password: user.password,
        enrolled: user.enrolled,
      },
    );
  };

  const handleNameChangePassword = async (event) => {
    setUser(
      {
        username: user.username,
        password: event.target.value,
        enrolled: user.enrolled,
      },
    );
  };

  const handleAsync = (event) => {
    event.preventDefault();
    // handleNameChangePassword(event).then(handleNameChange(event)).then(handleSubmit(event));
    handleSubmit(event);
  };

  const goToCreate = () => {
    navigate('/create');
  };

  return (
    <div className="Login">
      <p className="header">Research Match</p>
      <form onSubmit={handleAsync}>
        <label className="login-label" htmlFor="username">
          <div className="username-wrapper">
            <p className="field-label">USERNAME</p>
            <input
              className="input-field"
              type="text"
              id="username"
              onChange={handleNameChange}
            />
            <p className="field-label">PASSWORD</p>
            <input
              className="input-field"
              type="text"
              id="password"
              onChange={handleNameChangePassword}
            />
            <span className="error-message">{error.message}</span>
          </div>
          <input className="button" type="submit" value="SUBMIT" />
          <div className="spacer" />
          <button className="link" type="button" onClick={goToCreate}>New user sign-up</button>
        </label>
      </form>
    </div>
  );
}

export default Login;
