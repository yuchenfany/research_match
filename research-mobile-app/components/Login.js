/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React, { useState } from 'react';
import bcrypt from 'bcryptjs';
import ParticipantHome from './ParticipantHome';
import ParticipantStudies from './ParticipantStudies';
import Study from './Study'
import ResearcherHome from './ResearcherHome';
import ParticipantEdit from './ParticipantEdit';
import AddStudy from './AddStudy';
import ResearcherStudy from './ResearcherStudy';
import EditStudy from './EditStudy';
import DeleteAccount from './DeleteAccount';

import { NavigationContainer } from '@react-navigation/native';
import  { createNativeStackNavigator } from '@react-navigation/native-stack';


function Login({ navigation }) {
  const [study, setStudy] = useState({
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  });
  const [user, setUser] = useState({
    username: 'participant',
    password: 'participantpass',
    enrolled: [],
    phys: [],
    psych: [],
    med: [],
    studies: [],
  });
  const [error, setError] = useState({ message: '' });
  // const [samePassword, setSamePassword] = useState(0);

  async function handleSubmit(event) {
    console.log(user.username);
    console.log(user.password);

    if (user.username.length === 0 && user.password.length === 0) {
      setError({ message: 'Please enter your login credentials' });
      event.preventDefault();
      return;
    }
    if (user.username.length === 0) {
      setError({ message: 'Please enter your username' });
      event.preventDefault();
      return;
    }
    if (user.password.length === 0) {
      setError({ message: 'Please enter your password' });
      event.preventDefault();
      return;
    }

    const data = await fetch(`http://localhost:5000/record/${user.username}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    const json = await data.json();
    console.log(json);

    // verification checks of username & password
    if (json === null) {
      setError({ message: 'User does not exist' });
      event.preventDefault();
    //  } else if (bcrypt.compareSync(user.password, json.password)) { : CHANGE LATER 
    } else if (bcrypt.compareSync(user.password, json.password)) {
      if (json.type === 0) {
        // makes sure all fields are available in home
        await setUser({
          username: json.username,
          password: json.password,
          enrolled: json.enrolled,
          age: json.age,
          heightFeet: json.heightFeet,
          heightInches: json.heightInches,
          weight: json.weight,
          sex: json.sex,
          gender: json.gender,
          allergies: json.allergies,
          phys: json.phys,
          psych: json.psych,
          med: json.med,
          type: json.type,
        });

        console.log(user);
        console.log('JSON USER LOGIN');
        console.log(json.age);
        console.log(json.heightFeet);

        navigation.navigate('ParticipantHome', {
          user: user,
          setUser: setUser,
        });
      } else if (json.type === 1) {
        // makes sure all fields are available in home
        await setUser({
          username: json.username,
          password: json.password,
          name: json.name,
          organization: json.organization,
          studies: json.studies,
          type: json.type,
          title: json.title,
        });
        console.log(user);
        console.log(json.type);
        console.log(user.type);
        navigation.navigate('ResearcherHome', {
          user: user,
          setUser: setUser, 
          setStudy: setStudy
        });
        //navigate('/researcher-home');
      }
    } else {
      console.log('INCORRECT PASSWORD');
      setError({ message: 'Incorrect password' });
      event.preventDefault();
    }
  }

  // update username as it's being entered
  const handleNameChange = async (event) => {
    setUser(
      {
        username: event.target.value,
        password: user.password,
        enrolled: user.enrolled,
      },
    );
  };

  const handleNameChangePassword = async (event) => {
    setUser(
      {
        username: user.username,
        password: event.target.value,
        enrolled: user.enrolled,
      },
    );
  };

  const handleAsync = (event) => {
    event.preventDefault();
    // handleNameChangePassword(event).then(handleNameChange(event)).then(handleSubmit(event));
    handleSubmit(event);
  };

  return (
    <div className="Login">
      <p className="header">Research Match</p>
      <form onSubmit={handleAsync}>
        <label className="login-label" htmlFor="username">
          <div className="username-wrapper">
            <p className="field-label">USERNAME</p>
            <input
              className="input-field"
              type="text"
              id="username"
              onChange={handleNameChange}
            />
            <p className="field-label">PASSWORD</p>
            <input
              className="input-field"
              type="text"
              id="password"
              onChange={handleNameChangePassword}
            />
            <span className="error-message">{error.message}</span>
          </div>
          <input className="button" type="submit" value="SUBMIT" />
          <div className="spacer" />
        </label>
      </form>
    </div>
  );
}

// the app component will configure the screens and the routes

// create a navigation stack --> refrences the thing
const Stack = createNativeStackNavigator();
export default function App() {
  return (
    <NavigationContainer independent={true}>
      <Stack.Navigator initialRouteName="Research Application">
        <Stack.Screen name="Research Application" component={Login} />
        <Stack.Screen name="ParticipantHome" component={ParticipantHome} />
        <Stack.Screen name="ParticipantStudies" component={ParticipantStudies} />
        <Stack.Screen name="DeleteAccount" component={DeleteAccount} />
        <Stack.Screen name="Study" component={Study} />
        <Stack.Screen name="ResearcherHome" component={ResearcherHome} />
        <Stack.Screen name="ParticipantEdit" component={ParticipantEdit} />
        <Stack.Screen name="AddStudy" component={AddStudy} />
        <Stack.Screen name="ResearcherStudy" component={ResearcherStudy} />
        <Stack.Screen name="EditStudy" component={EditStudy} />
      </Stack.Navigator>
    </NavigationContainer>
  );

}
// export default Login;
