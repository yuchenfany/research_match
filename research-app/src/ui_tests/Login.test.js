/**
 * @jest-environment jsdom
 */

/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import React from 'react';
// import { jest } from '@jest/globals';

import Login from '../components/Login';

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

test('Login header', () => {
  const mockUser = {
    username: 'fakename', password: 'fakepass',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );
  const header = screen.getByText('Research Match');
  expect(header).toBeInTheDocument();
});

test('Login username', () => {
  const mockUser = {
    username: 'fakename', password: 'fakepass',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );
  const label = screen.getByText('USERNAME');
  expect(label).toBeInTheDocument();
});

test('Login submit', () => {
  const mockUser = {
    username: 'fakename', password: 'fakepass',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );
  fireEvent.click(screen.getByDisplayValue('SUBMIT'));
});

test('Login new user sign-up', () => {
  const mockUser = {
    username: 'fakename', password: 'fakepass',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );
  fireEvent.click(screen.getByText('New user sign-up'));
  expect(mockedUsedNavigate).toHaveBeenCalled();
});

test('Login sequence', () => {
  const mockUser = {
    username: '', password: '',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );

  const input = screen.getByPlaceholderText('username');
  fireEvent.change(input, { target: { value: 'user1' } });
  expect(input.value).toEqual('user1');
  expect(mockSetUser).toHaveBeenCalled();

  const input2 = screen.getByPlaceholderText('password');
  fireEvent.change(input2, { target: { value: 'user1pass' } });
  expect(input2.value).toEqual('user1pass');
  expect(mockSetUser).toHaveBeenCalled();

  fireEvent.click(screen.getByDisplayValue('SUBMIT'));
});

// test('Login sequence no password', () => {
//   const mockUser = {
//     username: '', password: '',
//   };
//   const mockSetUser = jest.fn();

//   render(
//     <Router>
//       <Login user={mockUser} setUser={mockSetUser} />
//     </Router>,
//   );

//   const input2 = screen.getByPlaceholderText('password');
//   fireEvent.change(input2, { target: { value: 'user1pass' } });
//   expect(input2.value).toEqual('user1pass');
//   expect(mockSetUser).toHaveBeenCalled();

//   fireEvent.click(screen.getByDisplayValue('SUBMIT'));
//   const error = renderer.create(
//     <span>Please enter your username</span>,
//   );
//   expect(error).toBeInTheDocument();
// });

test('snapshot test', () => {
  const mockUser = {
    username: 'fakename', password: 'fakepass',
  };
  const mockSetUser = jest.fn();

  const component = renderer.create(
    <Router>
      <Login user={mockUser} setUser={mockSetUser} />
    </Router>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
