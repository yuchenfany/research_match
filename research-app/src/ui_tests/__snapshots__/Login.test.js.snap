// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="Login"
>
  <p
    className="header"
  >
    Research Match
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      className="login-label"
      htmlFor="username"
    >
      <div
        className="username-wrapper"
      >
        <p
          className="field-label"
        >
          USERNAME
        </p>
        <input
          className="input-field"
          id="username"
          onChange={[Function]}
          placeholder="username"
          type="text"
        />
        <p
          className="field-label"
        >
          PASSWORD
        </p>
        <input
          className="input-field"
          id="password"
          onChange={[Function]}
          placeholder="password"
          type="text"
        />
        <span
          className="error-message"
        >
          
        </span>
      </div>
      <input
        className="button"
        type="submit"
        value="SUBMIT"
      />
      <div
        className="spacer"
      />
      <button
        className="link"
        onClick={[Function]}
        type="button"
      >
        New user sign-up
      </button>
    </label>
  </form>
</div>
`;
